contracts/Governance.sol

```diff
@@ -0,0 +1,480 @@
+
+    ShareLock storage lock = lockShareMap[voter][shareContract];
+    Proposal storage proposal = proposals[proposalId];
+
+    for (uint256 i = 0; i < lock.votedProposalIds.length; i++) {
+      require(lock.votedProposalIds[i] != proposalId, "already voted");
+    }
+
+    lock.votedProposalIds.push(proposalId);
+
+    proposal.voteIds.push(votes.length - 1);
+    if (support) {
+      proposal.forVotingPower = proposal.forVotingPower + lock.votingPower;
+    } else {
+      proposal.againstVotingPower = proposal.againstVotingPower + lock.votingPower;
+    }
+
+    emit ProposalVoted(proposalId, voter, support, shareContract, lock.amount, lock.votingPower);
+  }
+
+  function _voteForPoll(address voter, uint256 pollId, bool support, address shareContract) internal {
+    require(pollState(pollId) == ProposalState.Active, "vote not active");
+
+    ShareLock storage lock = lockShareMap[voter][shareContract];
+    Poll storage poll = polls[pollId];
+
+    for (uint256 i = 0; i < lock.votedPollIds.length; i++) {
+      require(lock.votedPollIds[i] != pollId, "already voted");
+    }
+
+    lock.votedPollIds.push(pollId);
+
+    poll.voteIds.push(votes.length - 1);
+    if (support) {
+      poll.forVotingPower = poll.forVotingPower + lock.votingPower;
+    } else {
+      poll.againstVotingPower = poll.againstVotingPower + lock.votingPower;
+    }
+
+    emit PollVoted(pollId, voter, support, shareContract, lock.amount, lock.votingPower);
+  }
+
+  function _paramInit() internal {
+    if (votingPeriod == 0) {
+      votingPeriod = INIT_VOTING_PERIOD;
+    }
+    if (pollVotingPeriod == 0) {
+      pollVotingPeriod = INIT_POLL_VOTING_PERIOD;
+    }
+    if (executionDelay == 0) {
+      executionDelay = INIT_EXECUTION_DELAY;
+    }
+    if (executionExpiration == 0) {
+      executionExpiration = INIT_EXECUTION_EXPIRATION;
+    }
+    if (quorumVotingPower == 0) {
+      quorumVotingPower = INIT_QUORUM_VOTING_POWER;
+    }
+    if (pollSubmitThreshold == 0) {
+      pollSubmitThreshold = INIT_POLL_SUBMIT_THRESHOLD;
+    }
+    if (minExecutableSupportRate == 0) {
+      minExecutableSupportRate = INIT_MIN_EXECUTE_SUPPORT_RATE;
+    }
+  }
+
+  function proposalState(uint256 proposalId) public view returns (ProposalState) {
+    require(proposalId < proposalLength(), "invalid proposal id");
+    Proposal storage proposal = proposals[proposalId];
+
+    uint256 totalVotingPower = proposal.forVotingPower + proposal.againstVotingPower;
+    uint256 executionVotingPowerThreshold = totalVotingPower * minExecutableSupportRate / PROPOSAL_EXECUTE_SUPPORT_RATE_SCALE;
+
+    if (proposal.canceled) {
+      return ProposalState.Canceled;
+    } else if (block.timestamp <= proposal.startAt) {
+      return ProposalState.Pending;
+    } else if (block.timestamp <= proposal.endAt) {
+      return ProposalState.Active;
+    } else if (proposal.forVotingPower <= executionVotingPowerThreshold ||  totalVotingPower < quorumVotingPower) {
```

Comment:
I think it is better that `executionVotingPowerThreshold` and `quorumVotingPower` is part of `proposal`, otherwise, even after the voting period end, the status of a proposal is not stable.